### PR TITLE
[FEATURE][RFC] pre loader (fully-contained extensions)

### PR DESCRIPTION
**Pull request by plastikos**
### What does this PR do?

Minimizes code in the loader core and allows all code related to an extension sub-system to be contained in the subsystem.  The loader does discovery to find directories that provide extensions and how to load the extensions.
### What issues does this PR fix or reference?

None.
### Previous Behavior

`salt/loader.py` needed to be modified each time extensions changed.  It was very large and a hotspot.
### New Behavior

All code related to an extension sub-system is contained in the sub-system directory.  `salt/loader.py` is much smaller and only has core `LazyLoader` code (arguably it was moved to `salt/loader_core.py` at the same time.
### Tests written?

No

### CODE REVIEW

**Comment by plastikos** [comment]
@cachedout - this is something I toyed around with while finishing up the minion key store.  I'm not expecting it to be merged - mostly fodder for discussion of why this may or may not be useful.  Mostly my tiny, little mind has been troubled since the beginning of time about parts of salt being less modular than expected (``salt/loader.py`` being one of those).

It's possible that this is a complete abomination for any of a variety of reasons - go ahead and tell me what you dislike.

File: salt/auth/.loader.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+'''Pre-loader for "auth" sub-system'''
+
+
+from __future__ import absolute_import
+
+from salt import loader_core
+from salt import loader_pre
+
+
+@loader_pre.LoaderFunc
```
Comment (plastikos) [inline salt/auth/.loader.py]: All of the old ``salt.loader.SUBSYSTEM()`` functions that were in ``salt/loader.py`` were moved into their respective sub-directories (viz the functions that instantiate a ``LazyLoader`` for each sub-directory).

File: salt/loader.py [hunk 2]
```diff
@@ -1,1685 +1,35 @@
 # -*- coding: utf-8 -*-
```
Comment (plastikos) [inline salt/loader.py]: ``salt/loader.py`` was gutted with core functions moving to ``salt/loader_core.py``

File: salt/loader_core.py [hunk 3]
```diff
@@ -0,0 +1,941 @@
+# -*- coding: utf-8 -*-
```
Comment (plastikos) [inline salt/loader_core.py]: Here's what was left-over from ``salt/loader.py`` after all the ``LazyLoader`` invocations were moved into the sub-system directories to ``.loader.py`` files.

File: salt/loader_pre.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
```
Comment (plastikos) [inline salt/loader_pre.py]: This is the sub-system discovery functions (viz. the new stuff that was added)

File: salt/loader_core.py [hunk 5]
```diff
@@ -0,0 +1,941 @@
...
+    ret = {}
+    funcs._load_all()
+    if filter_name:
+        funcs = FilterDictWrapper(funcs, filter_name)
+    for key in funcs:
+        ret[key] = funcs[key]
+    return ret
+
+
+def module_dirs(
```
Comment (plastikos) [inline salt/loader_core.py]: I renamed this from ``_module_dirs`` to ``module_dirs`` to squelch pylint warnings - it also is used outside of the loader (in a few spots) and consequently the ``_`` prefix is questionable.

**Comment by cachedout** [comment]
The more I read through this, the more I like it actually. Makes the loader system far more extensible and the ability for subsystems to contain their own loader interface, so to speak, makes a lot of sense to me.

What say you, @thatch45 ?

**Comment by thatch45** [comment]
While I get the motivation, this does kind of freak me out. I am trying to decide what I think

**Comment by cachedout** [comment]
@plastikos Could you rebase this when you get a chance?

**Comment by plastikos** [comment]
@cachedout , @thatch45 , yeah, I'm like that - I like to do things that come from nowhere and freak people out.  It's fair to resist, but I'm interested in what the gut feelings are.  Mine are that I like extensible tools to be fully extensible.  I don't like a central file that requires editing for each extension.

I'll rebase it.

**Comment by plastikos** [comment]
Rebased to #35543.

**Pull request closed without merging**